### PR TITLE
Policy: Refactor: inline main:GetMinRelayFee() in main:AcceptToMemoryPool()

### PR DESCRIPTION
**Pull request by jtimon**
main::GetMinRelayFee() is only called from main::AcceptToMemoryPool().
By in-lining in we can make some code simplifications and test that no new money is being created before calling main::AreInputsStandard(), instead of doing it in https://github.com/bitcoin/bitcoin/blob/master/src/main.cpp#L1032 after all those free transaction special cases. 
Even removing the `nFees < 0` check would be safe since it is repeated in https://github.com/bitcoin/bitcoin/blob/master/src/main.cpp#L1073 via https://github.com/bitcoin/bitcoin/blob/master/src/main.cpp#L1471
But moving it up seems more DoS-safe.
I've done the refactor step by step to to help with review.

### CODE REVIEW

**Comment by jtimon** [comment]
Needed rebase

**Comment by jtimon** [comment]
@luke-jr you once said you would like GetMinRelayFee to become part of policy, I would prefer to destroy it (and then pick it up as part of a new CPolicy method) instead.
Do you like this? 
By the way, in case it wasn't clear, this is supposed to be a 1 commit PR, as said it is just separated for easier review.

**Comment by jtimon** [comment]
I bet it will take very long until I have to rebase this because the refactored code is way too ugly for anyone to want to touch it. Hopefully by the time this gets any attention, +47 −67 0-functional-change will be enough of an argument in favor of this single-commit (I know, I have to s/SQUASHME:/fixup! but as said that may never happen because this may never need rebase). I would love to hear @morcos 's opinion.

File: src/main.cpp [hunk 1]
```diff
@@ -1003,6 +974,13 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
         view.SetBackend(dummy);
         }
 
+        CAmount nValueOut = tx.GetValueOut();
+        CAmount nFees = nValueIn-nValueOut;
+        // Don't accept it if it can't get into a block
+        if (nFees < 0)
+            return state.DoS(0, error("%s: negative fees %s, %d", __func__, hash.ToString(), nFees),
+                             REJECT_INSUFFICIENTFEE, "insufficient fee");
```
Comment (jtimon) [inline src/main.cpp]: Probably this should be replaced with bad-txns-fee-negative to better keep track of the validation duplication.

**Comment by morcos** [comment]
I took a first pass at reviewing this.  I think if you're going to do a refactor here, you should improve the readability of the code.  Although I like this code better than the original in some ways, it feels like you over optimized it instead of clarifying it.

In any case, I don't feel strongly, if you get some concept ACK's, I'm happy to test it for you.

I don't think your second commit is right though, why do you think minRelayTxFee is in MoneyRange already?

**Comment by jtimon** [comment]
Well, I was hoping that by doing the stuff change by change it could become obvious for reviewers that this doesn't change anything functionally, but thanks for the testing offer (probably better to do it after some more review like you say).

> I don't think your second commit is right though, why do you think minRelayTxFee is in MoneyRange already?

You are right, I thought, ParseMoney had a moneyRange check. I should add one there or in https://github.com/bitcoin/bitcoin/blob/master/src/init.cpp#L817

The other place where minRelayTxFee is initialized is in https://github.com/bitcoin/bitcoin/blob/master/src/main.cpp#L67

I will fix that commit.

Re optimizing: that's just a nice side effect. The main goal is to reduce the amount of code before moving it to policy.

**Comment by jtimon** [comment]
Updated fixing the second commit. Also reword "SQUASHME:" with "fixup!" in the commit descriptions

**Comment by jtimon** [comment]
Rebased

**Pull request closed without merging**